tests/host_callback_to_tf_test.py

```diff
@@ -0,0 +1,256 @@
+  tf = None
+
+config.parse_flags_with_absl()
+FLAGS = config.FLAGS
+
+
+def call_tf_no_ad(tf_fun: Callable, arg, *, result_shape):
+  """The simplest implementation of calling to TF, without AD support.
+
+  We must use hcb.call because the TF invocation must happen outside the
+  JAX staged computation."""
+
+  def tf_to_numpy(t):
+    return t.numpy() if isinstance(t, tf.Tensor) else t
+
+  return hcb.call(lambda arg: tf.nest.map_structure(tf_to_numpy,
+                                                    tf_fun(arg)),
+                  arg, result_shape=result_shape)
+
+
+def call_tf_simple_ad(tf_fun: Callable, arg, *, result_shape):
+  """Calls a TensorFlow function with simple support for reverse AD.
+
+  Works only for 1st order AD and only for arguments and results being a single
+  ndarray (no pytrees). Functions whose name starts with "tf_" are TensorFlow
+  functions and must be called outside the JAX computation.
+  """
+
+  @api.custom_vjp
+  def make_call(arg):
+    """We wrap it all in `make_call` so that we can attach custom VJP."""
+    return call_tf_no_ad(tf_fun, arg, result_shape=result_shape)
+
+  # Define the fwd and bwd custom_vjp functions
+  def make_call_vjp_fwd(arg):
+    # Return the primal argument as the residual. Use `make_call` for the
+    # primal computation to enable higher-order AD.
+    return make_call(arg), arg
+
+  def make_call_vjp_bwd(res, ct_res):
+    arg = res  # residual is the primal argument
+
+    def tf_vjp_fun(arg_and_ct_res):
+      """Invoke TF gradient; used with hcb.call."""
+      arg, ct_res = arg_and_ct_res
+      arg_var = tf.Variable(arg)
+      with tf.GradientTape(persistent=True) as tape:
+        res = tf_fun(arg_var)
+
+      dres_darg = tape.gradient(res, sources=arg_var,
+                                unconnected_gradients=tf.UnconnectedGradients.ZERO)
+      return dres_darg * ct_res
+
+    return (call_tf_simple_ad(tf_vjp_fun, (arg, ct_res),
+                              result_shape=arg),)
+
+  make_call.defvjp(make_call_vjp_fwd, make_call_vjp_bwd)
+  return make_call(arg)
+
+
+def call_tf_full_ad(tf_fun: Callable, arg, *, result_shape):
+  """Calls a TensorFlow function with support for reverse AD.
+
+  Supports higher-order AD and pytree arguments.
+  """
+
+  @api.custom_vjp
+  def make_call(arg):
+    """We wrap it all in `make_call` so that we can attach custom VJP."""
+    return call_tf_no_ad(tf_fun, arg, result_shape=result_shape)
+
+  # Define the fwd and bwd custom_vjp functions
+  def make_call_vjp_fwd(arg):
+    return make_call(arg), arg  # Return the primal argument as the residual
+
+  def make_call_vjp_bwd(res, ct_res):
+    arg = res  # residual is the primal argument
+
+    def tf_vjp_fun(arg_and_ct_res):
+      """Invoke TF gradient; used with hcb.call."""
+      arg, ct_res = arg_and_ct_res
+
+      def make_var(a):
+        return a if isinstance(a, tf.Variable) else tf.Variable(a)
+
+      arg_var = tf.nest.map_structure(make_var, arg)
```

Comment:
I think another viable option would be to simply use `tf.nest.map_structure(tf.convert_to_tensor, arg)`. Tensors also work fine in GradientTape.

Reply:
Actually this does not seem to work for higher-order differentiation, some dependencies get lost and I end up with 0 gradients. Perhaps you can try to clean this up once I merge this branch?